[sw/dif_pwrmgr] Power Manager DIF Library Header

alphan opened the pull request:
This change introduces the header of the power manager DIF library that defines its public interface.

## Notes
- The specification for power manager is not published yet but you can find it in lowRISC/opentitan#3119.
- The number of wake-up and reset sources are yet to be finalized. This DIF library can currently support upto 32 but we expect to have a small number of sources for each, e.g. 3 to 5.
- Wake-up and reset sources don't have any names yet. Once the design is finalized, we plan to define them (auto-generated) in `hw/top_earlgrey/sw/autogen/top_earlgrey.h` (lowrisc/openTitan#3138).
- Thus, the interface for interacting with sources may (probably will) change based on the final design.
- Reset reasons are recorded by reset manager, not power manager.
- `ABORT` wake-up reason _may_ go away (depends on if we really need to protect any peripherals from an illegal low power entry).
- @tjaychen please let me know if I missed anything.

Signed-off-by: Alphan Ulusoy <alphan@google.com>

jon-flatley commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,497 @@
+  /**
+   * The call succeeded.
+   */
+  kDifPwrmgrConfigResultOK = 0,
```
Thoughts on using `kDifPwrmgrConfigResultOk = kDifPwrmgrOk` style throughout?

lenary replied:
I think this is what is supposed to be used.

alphan replied:
I agree, thanks for catching this!

jon-flatley commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,497 @@
+/**
+ * Enable or disable sources for a request type.
+ *
+ * Setting `new_sources[i]` to 1 enables source `i` for `req_type` requests.
```
This is somewhat confusing as `i` refers to a bit index in `new_sources`.

alphan replied:
Updated comment.

jon-flatley commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,497 @@
+/**
+ * Get sources for a request type.
+ *
+ * Source `i` is enabled for `req_type` requests if `cur_sources[i]` is 1.
```
Same as above, but especially confusing here since `cur_sources` is a pointer.

jon-flatley commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,497 @@
+ */
+DIF_WARN_UNUSED_RESULT
+dif_pwrmgr_result_t dif_pwrmgr_irq_get(dif_pwrmgr_t *pwrmgr,
+                                       dif_pwrmgr_irq_t irq, bool *state);
```
`bool` -> `dif_pwrmgr_toggle_t`

alphan replied:
This is not super important but the reason why I prefer `bool` over `dif_pwrmgr_toggle_t` in this case is that the state of an IRQ is not something you can control and also to avoid confusion with enabling/disabling interrupts.

silvestrst replied:
I personally agree with @jon-flatley on this one, I think the type is more descriptive in this case.

silvestrst replied:
Oh, wait, I guess this is the function to get the corresponding `PWRMGR.INTR_STATE` bit? In this case maybe a better name for the function would be `dif_pwrmgr_irq_pending_get`. I misinterpreted this function as "irq enabled get".

alphan replied:
This has been resolved during new DIF naming guidelines.

tjaychen commented on meson.build:
```diff
@@ -116,6 +116,7 @@ hw_ip_spi_device_reg_h = gen_hw_hdr.process('hw/ip/spi_device/data/spi_device.hj
 hw_ip_rv_timer_reg_h = gen_hw_hdr.process('hw/ip/rv_timer/data/rv_timer.hjson')
 hw_ip_uart_reg_h = gen_hw_hdr.process('hw/ip/uart/data/uart.hjson')
 hw_ip_usbdev_reg_h = gen_hw_hdr.process('hw/ip/usbdev/data/usbdev.hjson')
+hw_ip_pwrmgr_reg_h = gen_hw_hdr.process('hw/ip/pwrmgr/data/pwrmgr.hjson')
```
`pwrmgr` is among the happy modules that have become templated, so please point `hw/top_earlgrey/ip/pwrmgr/data/autogen/pwrmgr.hjson` instead

tjaychen commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,497 @@
+ */
+DIF_WARN_UNUSED_RESULT
+dif_pwrmgr_result_t dif_pwrmgr_low_power_enabled_set(
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_toggle_t new_state);
```
i'm just curious, would it be clearer if you enumerated all the low power states and what each state's values should be?

alphan replied:
Can you please elaborate? This function enables transition to low power on the next `WFI`, i.e. the _low power hint_ in the spec. `dif_pwrmgr_low_power_config_set` configures what should happen in low power state. These are two separate functions so that clients can use the same low power configuration as long as they want.

tjaychen replied:
let me clarify.  All the functional bits in there technically enable us to build combinations of different power modes.
I was just wondering if it made sense to enumerate the valid power modes for a design, say deep sleep and sleep, and in those structures hardcode what the values might be.

Although thinking about it more, I guess this really should be the responsibility of the code that is one layer above the DIF, and not the DIF itself right?

alphan replied:
> I guess this really should be the responsibility of the code that is one layer above the DIF, and not the DIF itself right?

Thanks for clarifying. I agree, and TBH I thought about the same thing and I think it will depend on the number possible configurations and how flexible we want this DIF to be. IMHO, leaving this to the clients is a reasonable approach but we can always revisit this if we want the DIF to support only a few fixed configurations.

sriyerg commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,497 @@
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_pwrmgr_result_t dif_pwrmgr_irq_test(dif_pwrmgr_t *pwrmgr,
```
Most other IPs use the `_irq_force` nomenclature.

alphan commented:
Thanks for your reviews @jon-flatley, @sriyerg, @tjaychen, @lenary! I implemented your comments in the last push. Looking forward to more feedback.

silvestrst commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,499 @@
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_pwrmgr_result_t dif_pwrmgr_irq_enabled_set(dif_pwrmgr_t *pwrmgr,
```
I found `PWRMGR.INTR_ENABLE` `wakeup` bit description `Enable interrupt when INTR_STATE.wakeup is set` slightly confusing. Is it essentially saying that when this is enabled and `INTR_STATE.wakeup` is set, then an interrupt will occur?

alphan replied:
Isn't this same as other devices? `INTR_STATE` bits are set when the relevant condition occurs while `INTR_ENABLE` acts as a mask that determines which ones should actually trigger an interrupt. Please take a look at GPIO and I2C hw spec as well.

silvestrst commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,499 @@
+    const dif_pwrmgr_t *pwrmgr, dif_pwrmgr_wakeup_reason_t *reason);
+
+/**
+ * Clear wake-up reason and requests recorded during last low power state.
```
I remember in the past reviews @mcy felt strongly about function descriptions having form of  "Clears..." instead of "Clear...".

silvestrst commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,499 @@
+   * Peripheral sources that requested wake-up since the last time recording has
+   * started.
+   */
+  uint32_t sources;
```
How can we found out which peripheral caused a wake-up? `PWRMGR.WAKE_INFO.REASONS` is a single bit. Maybe the documentation for that bit could also be improved.

alphan replied:
I didn't define an enum because the peripherals that will send those requests don't exist yet, hence the single bit in the hardware. I will add an enum, which we can update later, to make the API clearer.

silvestrst commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,499 @@
+} dif_pwrmgr_req_type_t;
+
+/**
+ * Lock sources for a request type.
```
Could you please explain what "source" is, ctrl + f and searching for "source" in https://docs.opentitan.org/hw/ip/pwrmgr/doc/index.html didn't help much.

alphan replied:
Thanks for pointing this out. I added a comment that hopefully makes things clearer.

silvestrst commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,499 @@
+/**
+ * Enable or disable sources for a request type.
+ *
+ * Setting bit `i` of `new_sources` to 1 enables source `i` for `req_type`
```
As per above, could you explain what the sources are?

silvestrst reviewed (commented):
Thank you @alphan , I have done the first pass review. Looking good so far, will do another iteration after you had chance to address these and other comments.

mcy commented on sw/device/lib/dif/dif_pwrmgr.c:
```diff
@@ -0,0 +1,9 @@
+
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+
+#include "sw/device/lib/common.h"
```
common.h is going away, right?

mcy commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,499 @@
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_warn_unused_result.h"
+#include <stddef.h>
+#include <stdint.h>
```
System includes before project includes. (Also, have we been putting the file comment before or after includes? We should check to make sure we've picked on that's consistent.)

mcy commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,499 @@
+ * it.
+ */
+typedef struct dif_pwrmgr {
+  mmio_region_t base_addr;
+  uint8_t num_wakeup_sources;
+  uint8_t num_reset_sources;
```
Consider just keeping around the `config` struct, so that you don't need to redefine this.

silvestrst commented:
I think it is worth convert this header API to the new DIF template by @mcy . It will avoid churn, as it would be unfortunate to implement this API just to do name refactoring on both `.c`and  `.h`.

alphan commented:
Last update:
- Updated definitions to comply with new naming guidelines.
- Addressed comments.
- Updated API to use bitmasks and removed some redundant functions.
- Added DIF checklist.

@gkelly or @silvestrst can you take a quick look and let me know if you have any concerns?

silvestrst reviewed (approved):
@alphan thanks!

I have couple minor documentation suggestions, otherwise looks good to me!

I think if there are any problems with the API, we will discover them when you will start implementing the functions, so I think it is good to be merged!

lenary commented on sw/device/lib/dif/dif_pwrmgr.c:
```diff
@@ -0,0 +1,9 @@
+
+#include "sw/device/lib/common.h"
+
+#include "pwrmgr_regs.h"  // Generated.
```
Can you add some kind of `#error "Power Manager DIF not implemented"` or the like? So if other people try to depend on this, they get a clearer error than just missing link-time dependencies on the function symbols?

alphan replied:
Good point, it may be a while before the implementation is done. Added `#error` in `dif_pwrmgr.c` and commented out the new definition in `meson.build` in order not to break the build (did not delete it because I will work on the implementation next, hope that's OK).

lenary replied:
SG!

alphan commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,528 @@
+typedef uint8_t dif_pwrmgr_domains_t;
+
+/**
+ * A wakeup request source.
+ *
+ * Note: This needs to be updated once the HW is finalized.
```
Good point, done!

alphan commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,528 @@
+} dif_pwrmgr_wakeup_request_source_t;
+
+/**
+ * A reset request source.
+ *
+ * Note: This needs to be updated once the HW is finalized.
```
Good point, done!

alphan commented on sw/device/lib/dif/dif_pwrmgr.h:
```diff
@@ -0,0 +1,528 @@
+} dif_pwrmgr_req_type_t;
+
+/**
+ * A domain that can be enabled or disabled by the power manager.
```
Good point, done!

alphan commented:
Last update: Incorporated @silvestrst's and @lenary's comments.

I will merge this once the CI is done.
